### ISSUE TITLE
Using RTCBot to stream to multiple browser connections?

### ISSUE THREAD

**Issue by beniroquai** [dkumor/rtcbot#20]
Agter managing to install everything correctly I try to "enhance" the great tutorials a bit. 
I would like to have an always active client, which listens to the server similarly presented in the  [4G tutorial](https://rtcbot.readthedocs.io/en/latest/examples/mobile/README.html). I want to connect to this stream through the browser and I would like to do the same for another browser window (From another IP). 

Is this possible with your current library? 
Thanks! :-)

**Comment by dkumor** [comment]
This is actually a tough question, because it gets into the details of WebRTC. The short answer is yes, you can, but it is probably not what you want - at least not directly.

So WebRTC is really a method for connecting two computers together directly without an intermediate server. If you have single point-to-point communication (like robot to browser directly), this is great. However, if you want to connect 2 browsers to one robot, the way to do this in the same way would be for both browsers to connect to the same robot, and the robot to send two video streams at the same time. This is probably a bad idea, because not only does sending the same video stream twice eat up 4G bandwidth which might be limited, but also because it takes processing power to encode the two video streams optimized for both clients separately, and robots usually have limited computation.

A better approach, and indeed the one that is usually used by video-conferencing software is for the robot to actually connect with video to a main central server on the internet, and for that server to split the video stream into 2, and send it on to the client. This can be done with RTCBot, by having the server program actually read the video itself, and then send it on to all clients that are connected to it. However, at that point, it might be easier to not use WebRTC, since there is a lot of encoding/decoding that is happening. It might be a better approach to encode a video stream on the robot (maybe even using raspberry pi's hardware h264 encoder), send it to the server, and have the server just split and send the stream to both browser clients without processing.

TL;DR: yes you can, but RTCBot, and even WebRTC are not the best tools for the job.

**Comment by beniroquai** [comment]
I see the problem here. Thank you very much for your detailed explanation! For this, we would need something like a Janus instance on a server to "spread" the video stream I guess? 

We have an open-source project running which aims to build cheap microscopes. It's called [UC2](https://github.com/bionanoimaging/UC2-GIT/). You can see it as stationary robots, where you want to move the sample and turn on/off the light - remotely. So far the local network runs on MQTT and we would like to bridge a basic GUI to a framework called [ImJoy](https://imjoy.io/). We are mostly there with the help of your beautiful code. If you are interested, it would be fun to collaborate! :-) 

I must admit that I'm not a programmer, though I problems to make it work properly.. Finally, it should look like in the [Hackathon we recently planned](https://github.com/bionanoimaging/UC2-Hackathon#challenge). 
From your description, it makes no sense to spread the video to multiple browser clients. That's totally fine. 

Based on this, I have another question:
How could I re-connect to a client once the browser window is closed? You mention that in your tutorial, but I have no clue how I would program that. Some while loop, but how would I detect a closed browser window? 

Thanks a lot!

**Comment by dkumor** [comment]
Yeah, a server that sends out a single pre-compressed video stream to multiple outputs might be what you want here. Depending on your use case, the server could be on the Pi, for example, you can get a reasonable stream using mjpeg: https://randomnerdtutorials.com/video-streaming-with-raspberry-pi-camera/ - there are several approaches that might be suited for your goals, depending on your requirements.

UC2 looks great! I'm very happy that you found rtcbot useful, it might be close to what you want for your hackathon goals, so I really hope it ends up working well! Unfortunately I'm spread a bit too thin at the moment to work on additional projects, but I am happy to continue answering questions about RTCBot!

Regarding reconnecting a client after browser is closed:
You would want to keep a list of connections that are active - for example, an extension of the video streaming tutorial would be:
```python

camera = PiCamera()

# Here is a list of connections
connections = []

# This sets up the connection
@routes.post("/connect")
async def connect(request):
    clientOffer = await request.json()
    # Create a new connection
    conn = RTCConnection()
    # Subscribe it to the video feed
    frameSubscription = camera.subscribe()
    conn.video.putSubscription(frameSubscription)
    # When the connection is closed, unsubscribe from video, and remove connection from list
    @conn.onClose
    def close():
        camera.unsubscribe(frameSubscription)
        connections.remove(conn)
    # Add it to the list of connections
    connections.append(conn)
    serverResponse = await conn.getLocalDescription(clientOffer)
    return web.json_response(serverResponse)

# On cleanup, close all connections
async def cleanup(app):
    for conn in connections[:]:
        await conn.close()
```

**Comment by beniroquai** [comment]
Wow! Thanks for your detailed reply! I really appreciate that! 
May I quickly ask you which tutorial-template would be the starting point to get it to work? I'm not very familiar with sockets in Python - I apologize! 
I'll try on my own, but perhaps there is a very simple solution already ;-) 

And also: I can totally understand the lack of time! I really appreciate how much time and effort you already put into this project! It's a very good extension to what we are doing here! 🔬 everywhere! 
Thanks!

**Comment by dkumor** [comment]
There was actually a bug in connection closing that is fixed now. If you update rtcbot  (`pip install --upgrade rtcbot`), the following should do what you are looking for (and also supports multiple concurrent connections):

```python
from aiohttp import web

routes = web.RouteTableDef()

from rtcbot import RTCConnection, getRTCBotJS, CVCamera

camera = CVCamera()

# Serve the RTCBot javascript library at /rtcbot.js
@routes.get("/rtcbot.js")
async def rtcbotjs(request):
    return web.Response(content_type="application/javascript", text=getRTCBotJS())


# Here is a list of connections
connections = []

# This sets up the connection
@routes.post("/connect")
async def connect(request):
    print("Creating Connection")
    clientOffer = await request.json()
    # Create a new connection
    conn = RTCConnection()
    # Subscribe it to the video feed
    frameSubscription = camera.subscribe()
    conn.video.putSubscription(frameSubscription)
    # When the connection is closed, unsubscribe from video, and remove connection from list
    @conn.onClose
    def close():
        print("Connection Closed")
        camera.unsubscribe(frameSubscription)
        connections.remove(conn)

    # Add it to the list of connections
    connections.append(conn)
    serverResponse = await conn.getLocalDescription(clientOffer)
    return web.json_response(serverResponse)


@routes.get("/")
async def index(request):
    return web.Response(
        content_type="text/html",
        text="""
    <html>
        <head>
            <title>RTCBot: Video</title>
            <script src="/rtcbot.js"></script>
        </head>
        <body style="text-align: center;padding-top: 30px;">
            <video autoplay playsinline controls></video> <audio autoplay></audio>
            <p>
            Open the browser's developer tools to see console messages (CTRL+SHIFT+C)
            </p>
            <script>
                var conn = new rtcbot.RTCConnection();
                conn.video.subscribe(function(stream) {
                    document.querySelector("video").srcObject = stream;
                });
                async function connect() {
                    let offer = await conn.getLocalDescription();
                    // POST the information to /connect
                    let response = await fetch("/connect", {
                        method: "POST",
                        cache: "no-cache",
                        body: JSON.stringify(offer)
                    });
                    await conn.setRemoteDescription(await response.json());
                    console.log("Ready!");
                }
                connect();
            </script>
        </body>
    </html>
    """,
    )


async def cleanup(app):
    for conn in connections[:]:
        await conn.close()
    camera.close()


app = web.Application()
app.add_routes(routes)
app.on_shutdown.append(cleanup)
web.run_app(app)
```
